File: design_patterns/creational/Singleton/singleton.cpp
class Singleton
{
public:
    static Singleton *getInstance()
    {
        if (!instance)
            instance = new Singleton();
        return instance;
    }

    virtual void operation();

protected:
    Singleton();

private:
    static Singleton *instance;
};

int main()
{ // Usage
    Singleton::getInstance()->operation();
    return 0;
}

File: design_patterns/creational/Singleton/java/Singleton.java
package creational.Singleton.java;

public class Singleton {

    public static Singleton getInstance() {
        if (instance == null)
            instance = new Singleton();
        return instance;
    }

    protected Singleton() {
    }

    public void operation() {
    }

    private static Singleton instance;

}
